### PR TITLE
Makes grasp sampling more robust for ball and cup

### PR DESCRIPTION
**Pull request by NishanthJKumar**
The main point of this PR is as the name suggests. Experimentation with running oracle revealed that we (1) always want to grasp the top part of the cup, (2) want to force a top-down grasp on the ball for high accuracy. Thus, we now also sample a rotation quaternion constraint in addition to a grasp pixel.

Also, we got a new cup that is a yellow-ring, so we had to change the prompt to recognize this. One huge benefit of this cup is that it's impossible for it to land on its side (it's always resting stably on the floor), so we can just sample the top pixel and always pick it up that way!

### CODE REVIEW

File: predicators/spot_utils/perception/object_specific_grasp_selection.py [hunk 1]
```diff
@@ -55,17 +58,34 @@ def _get_ball_grasp_pixel(rgbds: Dict[str, RGBDImageWithContext],
     except KeyError:
         raise ValueError(f"{ball_obj} not detected in {camera_name}")
     x1, y1, x2, y2 = seg_bb.bounding_box
-    return int((x1 + x2) / 2), int((y1 + y2) / 2)
+    pixel = int((x1 + x2) / 2), int((y1 + y2) / 2)
+    return (pixel[0], pixel[1])
```
Comment (tomsilver) [inline predicators/spot_utils/perception/object_specific_grasp_selection.py:62]: isn't `pixel` already a tuple?

File: predicators/spot_utils/perception/object_specific_grasp_selection.py [hunk 2]
```diff
@@ -55,17 +58,34 @@ def _get_ball_grasp_pixel(rgbds: Dict[str, RGBDImageWithContext],
...
+def _get_cup_grasp_pixel(rgbds: Dict[str, RGBDImageWithContext],
+                         artifacts: Dict[str, Any],
+                         camera_name: str) -> Tuple[int, int]:
+    del rgbds
+    detections = artifacts["language"]["object_id_to_img_detections"]
+    try:
+        seg_bb = detections[cup_obj][camera_name]
+    except KeyError:
+        raise ValueError(f"{cup_obj} not detected in {camera_name}")
+    # Select the first (topmost) pixel from the mask. This ensures we
```
Comment (tomsilver) [inline predicators/spot_utils/perception/object_specific_grasp_selection.py:74]: left-most and top-most

**Review by tomsilver — APPROVED** [review]
makes sense to me!

**Review by tsilver-bdai — APPROVED** [review]
oops wrong account

File: predicators/ground_truth_models/spot_env/nsrts.py [hunk 3]
```diff
@@ -113,18 +114,23 @@ def _pick_object_from_top_sampler(state: State, goal: Set[GroundAtom],
...
-                                hand_camera, rng)
-
-    return np.array(pixel)
+        grasp_pixel = get_grasp_pixel(rgbds, artifacts, target_detection_id,
+                                      hand_camera, rng)
+        rot_quat_tuple = (0.0, 0.0, 0.0, 0.0)
+        if target_obj.name == "ball":
+            rot_quat = math_helpers.Quat.from_pitch(np.pi / 2)
+            rot_quat_tuple = (rot_quat.w, rot_quat.x, rot_quat.y, rot_quat.z)
+            params_tuple = grasp_pixel + rot_quat_tuple
```
Comment (tsilver-bdai) [inline predicators/ground_truth_models/spot_env/nsrts.py:131]: oops there's a bug here that I missed until running -- `params_tuple` not defined. surprised linting didn't catch. I'll push the fix.

**Pull request merged**